staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,244 @@
+	return cache
+}
+
+// initExistCache returns the slice of fileInfo for each of the files
+// in the existing cache directory, as well as the total size of
+// all these files. Sorts the fileInfo slice by ascending
+// file modification time (oldest files first). Returns an error
+// if one occurred during filepath.Walk.
+func initExistingCache(cacheDir string) ([]fileInfo, uint64, error) {
+	fileInfos := []fileInfo{}
+	currentSize := uint64(0)
+	err := filepath.Walk(cacheDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		fileSize := uint64(info.Size())
+		currentSize += fileSize
+		fi := fileInfo{
+			name:    info.Name(),
+			size:    fileSize,
+			modTime: info.ModTime(),
+		}
+		fileInfos = append(fileInfos, fi)
+		return nil
+	})
+	sort.Sort(byFileModTime(fileInfos)) // Sort by ascending modTime
+	return fileInfos, currentSize, err
+}
+
+// fileInfo stores relevant information for files in the disk cache.
+// This structure stores the size of each file to determine the
+// size of the entire disk cache, and it stores the modTime to
+// determine the oldest file to evict from the cache when necessary.
+// The name is the filename, which should be an MD5 hash (see
+// keyToFilename()).
+type fileInfo struct {
+	name    string
+	size    uint64
+	modTime time.Time
+}
+
+// Sort cache fileInfos by ascending last modification time.
+type byFileModTime []fileInfo
+
+func (fi byFileModTime) Len() int {
+	return len(fi)
+}
+
+func (fi byFileModTime) Less(i, j int) bool {
+	return fi[i].modTime.Before(fi[j].modTime)
+}
+
+func (fi byFileModTime) Swap(i, j int) {
+	fi[i], fi[j] = fi[j], fi[i]
+}
+
+// fileInfoIndex returns the index in the slice where the fileInfo
+// exists, and true if found. Returns unused integer and false if not found.
+func (c *SizeLimitedDiskCache) fileInfoIndex(key string) (int, bool) {
+	for i := 0; i < len(c.fileInfos); i++ {
+		if key == c.fileInfos[i].name {
+			return i, true
+		}
+	}
+	return 0, false
+}
+
+// removeFileInfo removes the fileInfo struct from the c.fileInfos
+// slice for the item with the name equals key. Returns true if
+// the items was found and deleted; false otherwise. Updates the
+// cache current size.
+func (c *SizeLimitedDiskCache) removeFileInfo(key string) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	i, exists := c.fileInfoIndex(key)
+	if exists {
+		c.currentSize -= c.fileInfos[i].size
+		c.fileInfos = append(c.fileInfos[:i], c.fileInfos[i+1:]...)
+	}
+	return exists
+}
+
+// appendFileInfo adds the passed fileInfo struct to the end of
+// the c.fileInfos slice, assuming this file is more recent than
+// all others. Before appending, any duplicates (by "name") are
+// first removed. Ensures no duplicate fileInfos exist in c.fileInfos.
+// Updates the cache currentSize.
+func (c *SizeLimitedDiskCache) appendFileInfo(fi fileInfo) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
Yeah I'm skeptical of using mutexes here, since you're probably competing with other processes as much or more than multiple threads within a single process.

Reply:
The scope of the protection for the mutex was meant to be modest; only protect the `fileInfos` indexing and the associated cache size calculation. I do not believe the it will be possible to keep the `fileInfos` faithfully synchronized with the actual files on disk. As you point out, there will be other processes possibly modifying the cache. Attempting a file lock on the cache directory seems very heavy weight. Do you think we should go down this path?

Reply:
I'm genuinely not sure! I think we need to know what happens when we fight with other processes at the very least.

Reply:
As it is currently coded, disk eviction competition is handled by just stopping the disk evictions. Specifically, in `trimDiskCache` if we attempt to run `diskv.Erase` on a file that has already been deleted, we just stop the eviction process. It is assumed that the file metadata in `cache.fileInfos` may be out of sync with the file system, and we may end up evicting slightly too aggressively or too passively. The eviction or non-eviction of a couple files it is assumed will not be material.